### PR TITLE
sql: generate index recommendations

### LINKED ISSUE #83782: Save index recommendations per statement fingerprint id
Collect index recommendations per fingerprint id.

We don't want to collect recommendation for all executions, so only calculate if there is no recommendations calculated in-memory so far. Steps:

1. Statement is executed
2. Check if there is a recommendation in-memory (crdb_internal.cluster_statement_statistics)
2a. If there isn't one, set a new flag to `true` indicating it should be calculated, and update [here](https://github.com/cockroachdb/cockroach/blob/b7c45c1631fbbaf908ce8919b8e23057feb2e8ca/pkg/sql/plan_opt.go#L558-L563) to also check this flag, calculate the recommendation and save in-memory
2b. If there is one, keep the flag as `false`
3. When the data in-memory flushed, replace the value persisted, with the latest value from memory

This solution keep the recommendations updated, and also doesn't generate a problem by calculating on every execution.

Challenge: since I'll calculate the rec the first time is added to in-memory, this means after a flush, there will be a spike on calculation, since everything will be new in-memory. Need to find a way to spread the calculation on the 10min period

Jira issue: CRDB-17261

### PR DESCRIPTION
**Pull request by maryliag**
We want to collect index recommendations
per statement and save it to the statement_statistics table.
To accomplish this, a new cache map was created.
The key for the map is a combination of the
statement fingerprint, the database and the plan hash,
since those are the parameters that would affect
an index recommendation (from the parameters we use
as keys for the statement fingerprint ID at least).

This cache has a limit of unique recommendations,
limited by a new cluster setting called
`sql.metrics.statement_details.max_mem_reported_idx_recommendations`
with a default value of 100k.
If this limit is reached, we delete entries that
had their last update made more than 24h ago.

A new recommendations is generated if has been more
than 1h since the last recommendation (saved on cache)
and it has at least a few executions (so we handle the case
where we have just one execution of several different
fingerprints, and is not worth calculating recommendations
for them).

The recommendations are saved as a list, with each
recommendations being "type : sql command", e.g.
`{"creation : CREATE INDEX ON t2 (i) STORING (k);"}`
and
`{"replacement : CREATE UNIQUE INDEX ON t1 (i) STORING (k);
DROP INDEX t1@existing_t1_i;"}`

Fixes #83782

---
Different approaches tested:
| branch  | Cache | Min Execution Count | Clear Cache|
| ------------- | ------------- | ------------- | ------------- |
| idx-rec-cache-cnt | Yes  | 1 | Older than 24h |
| idx-rec-large-cnt | Yes  | 10 | Older than 24h |
| idx-rec-random | Yes  | 1 | Random 1000 |
| idx-rec-no-cnt | Yes  | 0 | Older than 24h |
| idx-rec-no-cache | No  | N/A | N/A |

Min Execution Count is the minimum value of execution the statement must have so we decide to generate recommendation

Running TPC C on all branches, results:
 | branch | elapsed | tpmC | efc | avg(ms) | p50(ms) | p90(ms) | p95(ms) | p99(ms) | pMax(ms)|
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
| main | 4200.1s | 12.4 | 96.8%  | 19.8 | 18.9 | 25.2 | 31.5 | 46.1 | 302.0 | 
| idx-rec-cache-cnt | 3941.0s |12.6 | 98.3% | 18.8 | 18.9 | 23.1 | 25.2 | 37.7 | 52.4 | 
| idx-rec-large-cnt | 4138.8s | 12.3 | 95.7% | 18.0 | 17.8 | 22.0 | 24.1 | 39.8 | 130.0 | 
| idx-rec-random | 4321.9s | 12.6 | 98.1% | 19.3 | 18.9 | 24.1 | 27.3 | 39.8 | 209.7 | 
| idx-rec-no-cnt |  4200.2s | 12.4 | 96.4% | 19.3 | 18.9 | 24.1 | 26.2 | 44.0 | 67.1 | 
| idx-rec-no-cache | 4199.9s | 12.5 | 97.4% | 31.1 | 27.3 | 52.4 | 60.8 | 83.9 | 151.0 | 

The most significant change is when there is no cache, confirming the cache is necessary.  
The decision on keeping or removing the execution count check wasn't clear, so I decided to keep the count with a value of 5 and run another workload.

Running YSCB A:

 | branch | elapsed | errors | ops(total) | ops/sec(cum) | avg(ms) | p50(ms) | p95(ms) | p99(ms) | pMax(ms)|
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
| main | 4200.1s | 0 | 33334570 | 7936.6 | 4.0 | 2.9 | 11.0 | 18.9 | 436.2 |
| idx-rec-cache-cnt | 4012.4s | 0 | 20153193 | 5022.7 | 6.4 | 3.8 | 19.9 | 37.7 | 1879.0 | 
| idx-rec-no-cnt | 4200.1s | 0 | 29112702 | 6931.4 | 4.6 | 3.1 | 13.1 | 23.1 | 1342.2 | 
| idx-rec-no-cache | 4200.1s | 0 | 27216030 | 6479.8 | 4.9 | 3.4 | 13.6 | 23.1 | 872.4 |

Keeping the execution count has a lower performance when there is more contention, but since we have customers that generate several similar fingerprints by using different columns order on selects etc, we want to avoid creating recommendations for those cases, since it would be almost as having no cache. 
So I created another option (branch idx-rec-cache), where we do have a count, but once the count reaches the min count, we don't need to update it anymore (helping with contention on the cache itself). 
Running workloads with those options again:

TPC C
 | branch | elapsed | tpmC | efc | avg(ms) | p50(ms) | p90(ms) | p95(ms) | p99(ms) | pMax(ms)|
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
| main | 4200.1s | 12.6 | 98.0% | 19.0 | 18.9 | 22.0 | 23.1 | 27.3 | 62.9 |
| idx-rec-cache | 4200.0s | 12.5 | 97.0% | 19.3 | 19.9 | 23.1 | 24.1 | 31.5 | 62.9 |
| idx-rec-cnt | 4200.1s | 12.6 | 97.9% |  20.0 | 19.9 | 24.1 | 26.2 | 39.8 | 79.7 |
| idx-rec-no-cache | 4200.3s | 12.3 | 96.0% | 26.3 | 26.2 | 29.4 | 31.5 | 46.1 | 75.5 |

YSCB A
 | branch | elapsed | errors | ops(total) | ops/sec(cum) | avg(ms) | p50(ms) | p95(ms) | p99(ms) | pMax(ms)|
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
| main | 4200.0s  | 0 | 27374582 | 6517.8 | 4.9 | 3.4 | 13.1 | 28.3 | 872.4 |
| idx-rec-cache | 4200.0s | 0 | 26575359 | 6327.4 | 5.1 | 3.5 | 13.6 | 29.4 | 906.0 |
| idx-rec-cnt | 4199.9s | 0 | 27459203 | 6538.1 | 4.9 | 3.4 | 13.1 | 28.3 | 704.6 |
| idx-rec-no-cache | 4199.9s | 0 | 24826078 | 5911.1 | 5.4 | 3.7 | 14.2 | 30.4 | 738.2 |

---

Release note (sql change): Index recommendations are generated
for statements. New recommendations are generated every hour and
available on `system.statement_statistics` and
`crdb_internal.statement_statistics`.
New cluster setting with default value of 100k
sql.metrics.statement_details.max_mem_reported_idx_recommendations

### CODE REVIEW

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+	recommendations []string
+	executionCount  int64
+}
+
+type IndexRecCache struct {
+	st *cluster.Settings
+
+	// uniqueIndexRecInfoLimit is the limit on number of unique index
+	// recommendations info we can store in memory.
+	uniqueIndexRecInfoLimit *settings.IntSetting
```
Comment (ajwerner) [inline pkg/sql/idxrecommendations/idx_recommendations_cache.go:44]: generally settings are global variables -- it's pretty rare to want to have a struct that is parameterized over both the setting itself and the cluster settings object. A setting is really just a name and an offset.

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+		recommendations: recommendations,
+		executionCount:  execCount,
+	}
+}
+
+// ClearOlderIndexRecommendationsCache clear entries that was last updated
+// more than a day ago. Returns the total deleted entries.
+func (idxRec *IndexRecCache) ClearOlderIndexRecommendationsCache() int {
+	idxRec.mu.RLock()
+	defer idxRec.mu.RUnlock()
```
Comment (ajwerner) [inline pkg/sql/idxrecommendations/idx_recommendations_cache.go:188]: this needs to be locked exclusively

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	recInfo, found := idxRec.mu.idxRecommendations[key]
+
+	return recInfo, found
+}
+
+func (idxRec *IndexRecCache) setIndexRecommendations(
+	key indexRecKey, time time.Time, recommendations []string, execCount int64,
+) {
+	idxRec.mu.RLock()
+	defer idxRec.mu.RUnlock()
```
Comment (ajwerner) [inline pkg/sql/idxrecommendations/idx_recommendations_cache.go:155]: this needs to be locked exclusively

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+	idxRec.mu.idxRecommendations[key] = indexRecInfo{
+		lastGeneratedTs: time,
+		recommendations: recommendations,
+		executionCount:  execCount,
+	}
+}
+
+// ClearOlderIndexRecommendationsCache clear entries that was last updated
+// more than a day ago. Returns the total deleted entries.
+func (idxRec *IndexRecCache) ClearOlderIndexRecommendationsCache() int {
```
Comment (ajwerner) [inline pkg/sql/idxrecommendations/idx_recommendations_cache.go:186]: This is a very wordy method name. Also, does is need to be exported?

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+			atomic.AddInt64(idxRec.atomic.uniqueIndexRecInfo, int64(1))
+
+		// If we have exceeded limit of unique index recommendations try to delete older data.
+		if incrementedCount > limit {
+			deleted := idxRec.ClearOlderIndexRecommendationsCache()
+			// Abort if no entries were deleted.
+			if deleted == 0 {
+				atomic.AddInt64(idxRec.atomic.uniqueIndexRecInfo, -int64(1))
+				return
+			}
```
Comment (ajwerner) [inline pkg/sql/idxrecommendations/idx_recommendations_cache.go:173]: You're already doing the decrement inside the call to `ClearOlderIndexRecommnedationsCache`.

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+	return recInfo, found
+}
+
+func (idxRec *IndexRecCache) setIndexRecommendations(
+	key indexRecKey, time time.Time, recommendations []string, execCount int64,
+) {
+	idxRec.mu.RLock()
+	defer idxRec.mu.RUnlock()
+
+	_, found := idxRec.getIndexRecommendation(key)
```
Comment (ajwerner) [inline pkg/sql/idxrecommendations/idx_recommendations_cache.go:157]: consider returning early if you found it? A way I've seen this done elsewhere is to have a getOrCreate method which attempts to do the get and then re-locks exclusively and attempts to do the create assuming there was not a race.

**Review by ajwerner — COMMENTED** [review]
This is a drive-by review. Consider spelling out the interface for the cache a bit more clearly up-front in the sql package and then perhaps write some testing with a simpler implementation that, for example, just has a big ol' mutex on top of it and then iterate to a more optimized version. Generally best to write some micro-benchmarks with some concurrency to ensure that the optimizations are worth it.

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go (hunk 1 above)
Comment (maryliag) [reply pkg/sql/idxrecommendations/idx_recommendations_cache.go]: Done

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go (hunk 2 above)
Comment (maryliag) [reply pkg/sql/idxrecommendations/idx_recommendations_cache.go]: Fixed

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go (hunk 3 above)
Comment (maryliag) [reply pkg/sql/idxrecommendations/idx_recommendations_cache.go]: Fixed

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go (hunk 4 above)
Comment (maryliag) [reply pkg/sql/idxrecommendations/idx_recommendations_cache.go]: Initially I was considering also calling this function in another location periodically to help with the cleanup, but decided to keep it just here. Renamed to be local and smaller

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go (hunk 5 above)
Comment (maryliag) [reply pkg/sql/idxrecommendations/idx_recommendations_cache.go]: This is the case where the clean up didn't delete anything, meaning we can't add the new entry, so this reverses the value and return the function

File: pkg/sql/idxrecommendations/idx_recommendations_cache.go (hunk 6 above)
Comment (maryliag) [reply pkg/sql/idxrecommendations/idx_recommendations_cache.go]: I can't really return early, this is just looking if the entry exists, then I need to do the actual set of the new value. I did like the idea of the getOrCreate so made some updates to use that (here and in the other functions too)

**Review by j82w — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85343)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @j82w, and @maryliag)

___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 177 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8z6xdQ-EG20U0QRDVn:-N8z6xdQ-EG20U0QRDVo:b1hey5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L177)):*
> ```Go
> 		atomic.AddInt64(&idxRec.atomic.uniqueIndexRecInfo, int64(1))
> 
> 	if incrementedCount > limit {
> ```

If this hits the limit all the calls will go into the clearOldIdxRecommendations() which takes a lock. This will likely cause lock contention. To avoid the lock contention it might be better to track what the oldest recommendation is and skip the lock if there are none that are old. The other option I can think of it is track when it ran last and let it only run every x minutes.
___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8yzmrNC8wuzriv4mOB:-N8yzmrNC8wuzriv4mOC:bsqvfee) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L179)):*
> ```Go
> 	if incrementedCount > limit {
> 		// If we have exceeded limit of unique index recommendations try to delete older data.
> 		deleted := idxRec.clearOldIdxRecommendations()
> ```

Should this check the count again instead of looking at deleted to avoid possible race condition?



_Code snippet:_
```Go
Thread A -> add -> check -> clearOldIdxRecommendations() returns N so continues to add index.
       Thread B -> add -> check -> clearOldIdxRecommendations() returns 0 because it just cleaned a moment before by Thread A.
```
___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 191 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8z4bUX4QeqGR1NOsF4:-N8z4bUX4QeqGR1NOsF5:b-42lxcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L191)):*
> ```Go
> 	// For a new entry, we want the lastGeneratedTs to be in the past, in case we reach
> 	// the execution count, we should generate new recommendations.
> 	recInfo = indexRecInfo{
> ```

Would it be better to have a separate create method to avoid creating a default object?
___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 192 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8z2nSj3SPrlFQB8xLQ:-N8z2nSj3SPrlFQB8xLR:b-qjc6xb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L192)):*
> ```Go
> 	// the execution count, we should generate new recommendations.
> 	recInfo = indexRecInfo{
> 		lastGeneratedTs: timeutil.Now().Add(-time.Hour),
> ```

Why not use the default time `time.Time{}`?
___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 196 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8yyoZ_4oAHO0KjegjQ:-N8yyoZ_4oAHO0KjegjR:b-v27l1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L196)):*
> ```Go
> 		executionCount:  0,
> 	}
> 	idxRec.mu.idxRecommendations[key] = recInfo
> ```

I don't see a lock between the read above and this create. This allows a possible race condition where the original get returns false for the same key across multiple threads. They each then will come down and take the idxRec lock one by one and replace the recInfo just returned by the previous thread.

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85343)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @j82w)

___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 177 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8z6xdQ-EG20U0QRDVn:-N8zGlhg1Sb4H4UJOuNk:b1etcbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L177)):*
<details><summary><i>Previously, j82w wrote…</i></summary><blockquote>

If this hits the limit all the calls will go into the clearOldIdxRecommendations() which takes a lock. This will likely cause lock contention. To avoid the lock contention it might be better to track what the oldest recommendation is and skip the lock if there are none that are old. The other option I can think of it is track when it ran last and let it only run every x minutes.
</blockquote></details>

Done: decided to go with the approach of last cleanup ts.
___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8yzmrNC8wuzriv4mOB:-N8zGpvE-OQO_4kqe9eh:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L179)):*
<details><summary><i>Previously, j82w wrote…</i></summary><blockquote>

Should this check the count again instead of looking at deleted to avoid possible race condition?
</blockquote></details>

Done
___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 191 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8z4bUX4QeqGR1NOsF4:-N8zHlgTAony7nYJug6X:b-m8d85z) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L191)):*
<details><summary><i>Previously, j82w wrote…</i></summary><blockquote>

Would it be better to have a separate create method to avoid creating a default object?
</blockquote></details>

Sorry, I don't follow... this is the method to create. Do you mean one that just returns:
```
indexRecInfo{
		lastGeneratedTs: timeutil.Now().Add(-time.Hour),
		recommendations: []string{},
		executionCount:  0,
	}
```
___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 192 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8z2nSj3SPrlFQB8xLQ:-N8zIRkk7gHpwfqRriXR:bdw69b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L192)):*
<details><summary><i>Previously, j82w wrote…</i></summary><blockquote>

Why not use the default time `time.Time{}`?
</blockquote></details>

Imagine I have a query that runs hourly, this means on hour 5 I will get a recommendation for it, but if on hour 3 we reached the cache limit I'll try to delete data older than 24h, this means:
with this approach: this entry will continue on the cache and will eventually get the recommendation
with the approach of using `time.Time{}` this entry will get deleted and I won't get recommendations for it
___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 196 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8yyoZ_4oAHO0KjegjQ:-N8zJddh9S8PLzvG9bK_:b3uem9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L196)):*
<details><summary><i>Previously, j82w wrote…</i></summary><blockquote>

I don't see a lock between the read above and this create. This allows a possible race condition where the original get returns false for the same key across multiple threads. They each then will come down and take the idxRec lock one by one and replace the recInfo just returned by the previous thread. 
</blockquote></details>

Currently I do a RLock to get the value, if it returns false, I do a Lock to create a value, which could have a race condition, but the alternative would be starting a full lock, getting the value and if false creating the new entry, and then unlocking. The problem with that approach is that because we have a lot more gets than create, you could create more contention if we had to do a full lock for all gets. If the alternative is to maybe have a race condition, I think it's okay for this case, since the first few executions wouldn't generate recommendation anyway.

**Review by j82w — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85343)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @j82w, and @maryliag)

___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 191 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8z4bUX4QeqGR1NOsF4:-N91h6ozBfkglYzpb-Yl:b-gfh4dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L191)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

Sorry, I don't follow... this is the method to create. Do you mean one that just returns:
```
indexRecInfo{
		lastGeneratedTs: timeutil.Now().Add(-time.Hour),
		recommendations: []string{},
		executionCount:  0,
	}
```
</blockquote></details>

I was trying to think of a solution to avoid that ambiguity of it returning a new indexRecInfo vs one that is already exists with a recommendation. Looking it over again I think it is fine.
___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 192 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8z2nSj3SPrlFQB8xLQ:-N91lL_t3DpVM3d5KQMr:b-ykak1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L192)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

Imagine I have a query that runs hourly, this means on hour 5 I will get a recommendation for it, but if on hour 3 we reached the cache limit I'll try to delete data older than 24h, this means:
with this approach: this entry will continue on the cache and will eventually get the recommendation
with the approach of using `time.Time{}` this entry will get deleted and I won't get recommendations for it
</blockquote></details>

Good point, I didn't consider the clean up task. My only concern is if it the clean up window gets lowered to 1h for some reason this will break. 
___
*[`pkg/sql/idxrecommendations/idx_recommendations_cache.go` line 196 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N8yyoZ_4oAHO0KjegjQ:-N91e1vj73pkt091jaMr:busvlhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1abfd8a5acedaa29c7cc2bb0bb30b9c3d8d1cdd/pkg/sql/idxrecommendations/idx_recommendations_cache.go#L196)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

Currently I do a RLock to get the value, if it returns false, I do a Lock to create a value, which could have a race condition, but the alternative would be starting a full lock, getting the value and if false creating the new entry, and then unlocking. The problem with that approach is that because we have a lot more gets than create, you could create more contention if we had to do a full lock for all gets. If the alternative is to maybe have a race condition, I think it's okay for this case, since the first few executions wouldn't generate recommendation anyway.
</blockquote></details>

This is the solution I've seen implemented where it does another check after taking the full lock to verify another thread did not update it.



_Code snippet:_
```Go
	recInfo, found := idxRec.getIndexRecommendation(key)
	if found {
		return recInfo, true
	}
	
	...
	
	idxRec.mu.Lock()
	defer idxRec.mu.Unlock()
	
	recInfo, found = idxRec.mu.idxRecommendations[key]
	if found {
		return recInfo, true
	}
	
	recInfo = indexRecInfo{...}
	return recInfo, true
	
```

**Review by j82w — APPROVED** [review]
> Reviewed 13 of 16 files at r2, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85343)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @maryliag)

**Review by yuzefovich — COMMENTED** [review]
Changes in `sql` package LGTM, just a few nits.

> Reviewed 4 of 16 files at r2, 1 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-:-N92XhAf-03wL-WLrdtY:bil93j6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @maryliag)

___
*[`pkg/roachpb/app_stats.go` line 158 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N92X-G2-fR8dUOaP2T0:-N92X-G2-fR8dUOaP2T1:b-85bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4f2c416fdf0ea6e8e27e3961ac26255d7e8cf9f/pkg/roachpb/app_stats.go#L158)):*
> ```Go
> 	s.Nodes = util.CombineUniqueInt64(s.Nodes, other.Nodes)
> 	s.PlanGists = util.CombineUniqueString(s.PlanGists, other.PlanGists)
> 	s.IndexRecommendations = util.CombineUniqueString(s.IndexRecommendations, other.IndexRecommendations)
> ```

Why did this change?
___
*[`pkg/sql/instrumentation.go` line 669 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N92XKQ91bDcoCRx30kf:-N92XKQ91bDcoCRx30kg:b-3dpwy1) ([raw file](https://github.com/cockroachdb/cockroach/blob/31aa74151e65b74cdc5cb69c0a3e5e71945e7ca6/pkg/sql/instrumentation.go#L669)):*
> ```Go
> }
> 
> // SetIndexRecommendations check if we should generate a new index recommendation.
> ```

nit: `s/check/checks/` `s/use the value/uses the value/` `s/and update/and updates/`.
___
*[`pkg/sql/instrumentation.go` line 686 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N92XUwr3JuvQp4hgqT6:-N92XUwr3JuvQp4hgqT7:b-t5j1xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/31aa74151e65b74cdc5cb69c0a3e5e71945e7ca6/pkg/sql/instrumentation.go#L686)):*
> ```Go
> 	) {
> 		f := opc.optimizer.Factory()
> 		// EvalContext() has the context of the already closed span, so we need to update with the current context.
> ```

nit: `s/of the/with the`, also wrap at 80 characters.

**Comment by maryliag** [comment]
TFTR!
bors r+

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85343)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/roachpb/app_stats.go` line 158 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N92X-G2-fR8dUOaP2T0:-NFzfT-X4eda9LNqCWkr:bpz8zux) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4f2c416fdf0ea6e8e27e3961ac26255d7e8cf9f/pkg/roachpb/app_stats.go#L158)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Why did this change?
</blockquote></details>

@maryliag independently from the comment from two months ago I arrived at this code line again, and I'm still curious - why did this line change in this PR? :)

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85343)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/roachpb/app_stats.go` line 158 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85343#-N92X-G2-fR8dUOaP2T0:-N92c_Wy5BsyD294hId5:bjiap0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4f2c416fdf0ea6e8e27e3961ac26255d7e8cf9f/pkg/roachpb/app_stats.go#L158)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

@maryliag independently from the comment from two months ago I arrived at this code line again, and I'm still curious - why did this line change in this PR? :)
</blockquote></details>

(I opened here just to find that I did write the reply, but was still on draft and never got send :facepalm: sorry about that)

When I added this line on a previous PR I hadn't quite decided the format it would come from the index recommendation on cache, so it was just preparing to persist and I just used the default as combining like we have for everything else.  
Now I always want to update my stats with the latest recommendation I have. If I use the combine I might get some weird combination that no longer make sense. 

Imagine you run a statement and have 2 recommendations, the format of the `IndexRecommendations` is a string with all recommendations, e.g. `rec1; rec2;`, then you run again, but this time you get `rec3, rec 4;`, I only care about `rec3; rec4;`, so I overwrite what I had on the previous one, otherwise combining would give me `rec1; rec2; rec3; rec4;`. Maybe the confusion is coming from imagining this is combining a `rec1` and `rec2`, which is not the case.
Does that help?

**Comment by yuzefovich** [comment]
Yes, makes sense, thanks for explaining!

**Pull request merged**